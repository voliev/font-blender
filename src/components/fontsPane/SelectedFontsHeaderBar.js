import React from "react";
import classNames from "classnames";
import ExpandLessRounded from "@material-ui/icons/ExpandLessRounded";
import ExpandMoreRounded from "@material-ui/icons/ExpandMoreRounded";
import PropTypes from "prop-types";
import "./styles/SelectedFontsHeaderBar.css";

const SelectedFontsHeaderBar = ({
  folded,
  selectedFonts,
  headerBarClickHandler
}) => {
  const componentClasses = classNames({
    SelectedFontsHeaderBar: true,
    empty: selectedFonts.length === 0
  });

  return (
    <div className={componentClasses} onClick={headerBarClickHandler}>
      <p className="text">
        <span className="num">{selectedFonts.length}</span>
        <span>
          {" "}
          {selectedFonts.length === 1 ? "family" : "families"} selected
        </span>
      </p>
      <div className="icon">
        {folded ? <ExpandLessRounded /> : <ExpandMoreRounded />}
      </div>
    </div>
  );
};

SelectedFontsHeaderBar.propTypes = {
  folded: PropTypes.bool.isRequired,
  selectedFonts: PropTypes.arrayOf(PropTypes.string.isRequired).isRequired,
  headerBarClickHandler: PropTypes.func.isRequired
};

export default SelectedFontsHeaderBar;
