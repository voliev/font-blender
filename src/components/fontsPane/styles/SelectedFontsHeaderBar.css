.SelectedFontsHeaderBar {
  align-items: center;
  background: var(--blue);
  color: var(--white);
  cursor: pointer;
  display: flex;
  height: 3rem;
  justify-content: space-between;
  transition: background .3s, color .3s;
}

.SelectedFontsHeaderBar .text {
  align-items: center;
  display: flex;
  font-weight: 700;
  height: 100%;
  margin: 0 0 0 var(--space-lg);
}

.SelectedFontsHeaderBar .num {
  align-items: center;
  background: var(--white);
  border-radius: var(--border-radius-max);
  color: var(--blue);
  display: flex;
  height: 2rem;
  justify-content: center;
  margin-right: var(--space-xs);
  width: 2rem;
}

.SelectedFontsHeaderBar .icon {
  align-items: center;
  display: flex;
  height: 100%;
  justify-content: center;
  margin: 0 var(--space-lg) 0 0;
}

.SelectedFontsHeaderBar.empty {
  background: var(--dark-grey);
}

.SelectedFontsHeaderBar.empty .num {
  color: var(--dark-grey);
}
