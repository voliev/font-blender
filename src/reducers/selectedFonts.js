import undoable, { includeAction } from "redux-undo";
import {
  ADD_SELECTED_FONT,
  CLEAR_SELECTED_FONTS,
  REMOVE_FONT_FROM_SELECTED,
  UNDO_DESELECTED
} from "../constants";
import { getFonts } from "./fonts";

const initialState = [];

export const selectedFonts = (state = initialState, action) => {
  switch (action.type) {
    case ADD_SELECTED_FONT:
      return [...state, action.family];
    case REMOVE_FONT_FROM_SELECTED:
      return state.filter(family => family !== action.family);
    case CLEAR_SELECTED_FONTS:
      return [];
    default:
      return state;
  }
};

export default undoable(selectedFonts, {
  limit: 1,
  filter: includeAction([REMOVE_FONT_FROM_SELECTED, CLEAR_SELECTED_FONTS]),
  undoType: UNDO_DESELECTED
});

export const getSelectedFontFamilies = state => state.selectedFonts.present;
export const getSelectedFonts = state => {
  const fonts = getFonts(state);

  return getSelectedFontFamilies(state).reduce((selected, family) => {
    selected[family] = fonts[family];
    return selected;
  }, {});
};
export const isSelected = (state, family) =>
  state.selectedFonts.present.includes(family);
