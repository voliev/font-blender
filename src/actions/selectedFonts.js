import {
  ADD_SELECTED_FONT,
  CLEAR_SELECTED_FONTS,
  REMOVE_FONT_FROM_SELECTED,
  UNDO_DESELECTED
} from "../constants";
import { startFetchFont } from "./fonts";
import { addNotification, removeNotification } from "./notifications";
import { getFonts } from "../reducers/fonts";

export const addFontToSelected = family => ({
  type: ADD_SELECTED_FONT,
  family
});

export const removeFontFromSelected = family => ({
  type: REMOVE_FONT_FROM_SELECTED,
  family
});

export const clearSelectedFonts = () => ({
  type: CLEAR_SELECTED_FONTS
});

export const startAddFontToSelected = family => {
  return (dispatch, getState) => {
    const state = getState();
    const fonts = getFonts(state);
    const font = fonts[family];

    if (font.fetched) {
      // Add font to selected if it has been already fetchedselected list
      dispatch(addFontToSelected(family));
    } else {
      // Fetch font first than add it to selected
      return dispatch(startFetchFont(family))
        .then(() => dispatch(addFontToSelected(family)))
        .catch(() => {});
    }
  };
};

export const undoDeselected = () => ({
  type: UNDO_DESELECTED
});

export const startRemoveFontFromSelected = family => {
  return dispatch => {
    dispatch(removeFontFromSelected(family));

    const text = `${family} font family has been removed`;
    const group = "selectedFonts";
    // Add notification to display it with "undo" option
    dispatch(addNotification(text, group));
  };
};

export const startClearSelectedFonts = () => {
  return (dispatch, getState) => {
    const state = getState();
    const selectedFonts = [...state.selectedFonts.present];

    dispatch(clearSelectedFonts());

    const text =
      selectedFonts.length > 1
        ? `${selectedFonts.length} font families have been removed`
        : `${selectedFonts.length} font family have been removed`;
    const group = "selectedFonts";
    // Add notification to display it with "undo" option
    dispatch(addNotification(text, group));
  };
};

export const startUndoDeselected = () => {
  return dispatch => {
    // Undo deselection
    dispatch(undoDeselected());

    const group = "selectedFonts";

    // Remove notification to hide "undo" option
    dispatch(removeNotification(group));
  };
};
