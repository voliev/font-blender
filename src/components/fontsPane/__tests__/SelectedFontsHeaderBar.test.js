import React from "react";
import { mount, shallow } from "enzyme";
import SelectedFontsHeaderBar from "../SelectedFontsHeaderBar";
import ExpandLessRounded from "@material-ui/icons/ExpandLessRounded";
import ExpandMoreRounded from "@material-ui/icons/ExpandMoreRounded";
import selectedFonts from "../../../fixtures/selectedFonts";

const setup = (renderFn = shallow, propsOverrides = {}) => {
  const props = {
    folded: true,
    selectedFonts,
    headerBarClickHandler: jest.fn(),
    ...propsOverrides
  };

  return {
    ...props,
    wrapper: renderFn(<SelectedFontsHeaderBar {...props} />)
  };
};

afterEach(() => {
  jest.clearAllMocks();
});

it("should render without crashing", () => {
  const { wrapper } = setup();

  expect(wrapper).toMatchSnapshot();
});

it("should not crash if the list is empty", () => {
  const { wrapper } = setup(shallow, { selectedFonts: [] });

  expect(wrapper).toMatchSnapshot();
});

it("should receive corresponding props", () => {
  const { wrapper, headerBarClickHandler } = setup(mount);

  expect(typeof wrapper.prop("folded")).toBe("boolean");
  expect(wrapper.prop("folded")).toBeTruthy();

  expect(typeof wrapper.prop("selectedFonts")).toBe("object");
  expect(wrapper.prop("selectedFonts").hasOwnProperty("length")).toBeTruthy();
  expect(wrapper.prop("selectedFonts")).toEqual(selectedFonts);

  expect(typeof wrapper.prop("headerBarClickHandler")).toBe("function");
  expect(wrapper.prop("headerBarClickHandler")).toBe(headerBarClickHandler);
});

it("should set corresponding component classes", () => {
  const { wrapper } = setup(shallow, { selectedFonts: [] });

  expect(wrapper.hasClass("SelectedFontsHeaderBar")).toBeTruthy();
  expect(wrapper.hasClass("empty")).toBeTruthy();

  wrapper.setProps({ selectedFonts });

  expect(wrapper.hasClass("SelectedFontsHeaderBar")).toBeTruthy();
  expect(wrapper.hasClass("empty")).toBeFalsy();
});

it("should set corresponding icon", () => {
  const { wrapper } = setup(mount, { selectedFonts: [] });

  expect(wrapper.find(ExpandLessRounded).length).toBe(1);
  expect(wrapper.find(ExpandMoreRounded).length).toBe(0);

  wrapper.setProps({ selectedFonts });

  expect(wrapper.find(ExpandLessRounded).length).toBe(1);
  expect(wrapper.find(ExpandMoreRounded).length).toBe(0);

  wrapper.setProps({ folded: false });

  expect(wrapper.find(ExpandLessRounded).length).toBe(0);
  expect(wrapper.find(ExpandMoreRounded).length).toBe(1);

  wrapper.setProps({ folded: true });

  expect(wrapper.find(ExpandLessRounded).length).toBe(1);
  expect(wrapper.find(ExpandMoreRounded).length).toBe(0);
});

it("should set corresponding text", () => {
  const { wrapper } = setup(mount, { selectedFonts: [] });

  expect(wrapper.find(".text").text()).toBe("0 families selected");

  wrapper.setProps({ selectedFonts: selectedFonts.slice(0, 1) });

  expect(wrapper.find(".text").text()).toBe("1 family selected");

  wrapper.setProps({ selectedFonts });

  expect(wrapper.find(".text").text()).toBe(
    `${selectedFonts.length} families selected`
  );
});

it("should call handler on header click", () => {
  const { wrapper, headerBarClickHandler } = setup();

  wrapper.simulate("click");

  expect(headerBarClickHandler).toHaveBeenCalledTimes(1);

  wrapper.setProps({ selectedFonts: [] });
  wrapper.simulate("click");

  expect(headerBarClickHandler).toHaveBeenCalledTimes(2);
});
