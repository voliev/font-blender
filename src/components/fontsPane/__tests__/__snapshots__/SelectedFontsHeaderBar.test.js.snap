// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should not crash if the list is empty 1`] = `
<div
  className="SelectedFontsHeaderBar empty"
  onClick={[MockFunction]}
>
  <p
    className="text"
  >
    <span
      className="num"
    >
      0
    </span>
    <span>
       
      families
       selected
    </span>
  </p>
  <div
    className="icon"
  >
    <ExpandLessRoundedIcon />
  </div>
</div>
`;

exports[`should render without crashing 1`] = `
<div
  className="SelectedFontsHeaderBar"
  onClick={[MockFunction]}
>
  <p
    className="text"
  >
    <span
      className="num"
    >
      4
    </span>
    <span>
       
      families
       selected
    </span>
  </p>
  <div
    className="icon"
  >
    <ExpandLessRoundedIcon />
  </div>
</div>
`;
